### PR TITLE
Category Product filter optimized .

### PR DESCRIPTION
**Pull request by scchethu**
**getProductsRelatedFilterableAttributes()** takes more time to make page-load in the category tree, because of **n*n** loop.

[#768791975](https://github.com/bagisto/bagisto/issues/4369#issue-768791975) ,this issue optimized with better solution.
Please look at my PR.

### CODE REVIEW

**Comment by ghermans** [comment]
Thank you for this pull request @scchethu, we will review it the upcoming days.

File: packages/Webkul/Category/src/Models/Category.php [hunk 1]
```diff
@@ -66,8 +66,15 @@ public function getImageUrlAttribute()
      */
     public function filterableAttributes()
     {
+
+
         return $this->belongsToMany(AttributeProxy::modelClass(), 'category_filterable_attributes')->with(['options' => function($query) {
+            $product_ids = $this->products()->pluck('id');
+            $values = ProductAttributeValue::whereNotNull('integer_value')
```
Comment (devansh-webkul) [inline packages/Webkul/Category/src/Models/Category.php:73]: This class has not been imported. Please check the class namespace.

File: packages/Webkul/Category/src/Models/Category.php [hunk 2]
```diff
@@ -66,8 +66,15 @@ public function getImageUrlAttribute()
      */
     public function filterableAttributes()
     {
+
```
Comment (devansh-webkul) [inline packages/Webkul/Category/src/Models/Category.php:69]: 2 blank lines are added, please remove this,

**Review by devansh-webkul — CHANGES_REQUESTED** [review]
Hi @scchethu,

Please refer to these changes, however, I suggest this is not a solution. This will just prevent the first case i.e. when you load the category. Just try to filter color on that category also.

**Comment by devansh-webkul** [comment]
Hi, @ghermans and @scchethu,

There is already one more PR i.e. https://github.com/bagisto/bagisto/pull/4397 which is related to optimization.

File: packages/Webkul/Category/src/Models/Category.php (hunk 2 above)
Comment (scchethu) [reply packages/Webkul/Category/src/Models/Category.php:69]: Ok Wil resubmit with your changes.

**Comment by devansh-webkul** [comment]
Hi @scchethu,

I have 3500+ products in one category, now try to filter by color or some other attribute i.e.,

Screenshot

It is not optimized yet however, this pr will only load the main category page.

**Review by scchethu — COMMENTED** [review]
> Hi @scchethu,
> 
> I have 3500+ products in one category, now try to filter by color or some other attribute i.e.,
> 
> Screenshot
> 
> It is not optimized yet however, this pr will only load the main category page.

Ok, I will review my code again to fix the color filter as well.

**Comment by devansh-webkul** [comment]
Hi @scchethu,

This is not limited to a color filter, filter can be of any type like brands, price, etc.

**Comment by devansh-webkul** [comment]
Hi @scchethu,

Thank you for your effort, we really value each contribution.

Unfortunately, this PR needs to be closed because one of the internal developers already raised a PR.

Please check this PR i.e. https://github.com/bagisto/bagisto/pull/4397.

**Pull request closed without merging**